File: exercise/python/parsing_xml.py
#!/usr/bin/python
#-*- coding: utf-8 -*-

"""
Parse uniprot's xml to extract some information for the protein.
"""

import requests
from StringIO import StringIO
from lxml import etree

if __name__ == '__main__':
    # Load the information from uniprot
    xml = requests.get('http://www.uniprot.org/uniprot/Q42435.xml')
    # Remove the reference to the xml schema in the second line of the
    # file, it confuses lxml
    content = xml.content.split('\n')
    content[1] = '<uniprot>'
    xml_tree = etree.fromstring('\n'.join(content))

    # Retrieve the name of the protein
    names = []
    for name in xml_tree.xpath('entry/protein/recommendedName/fullName'):
        names.append(name.text)
    print 'Protein has for name         : %s' % ', '.join(names)

    # Retrieve the short name of the protein
    short_names = []
    for short_name in xml_tree.xpath('entry/name'):
        short_names.append(short_name.text)
    print 'Protein has for short name   : %s' % ', '.join(short_names)

    # Retrieve the list of pathways to which the protein is known to be
    # involved in
    pathways = []
    for annotation in xml_tree.xpath('entry/comment'):
        if annotation.attrib['type'] == 'pathway':
            for element in annotation.xpath('text'):
                pathways.append(element.text)
    print 'Protein has for pathways     : %s' % ', '.join(pathways)

File: exercise/perl/parsing_xml.pl
#!/usr/bin/perl

use strict;
use warnings;

=head1 DESCRIPTION

This example script retrieves the xml file from uniprot and extract
information from it.

=cut

use XML::Simple;
use LWP::Simple;
use Data::Dumper;

my $xs1 = XML::Simple->new();

# Retrieve the xml from the website
my $html = get("http://www.uniprot.org/uniprot/Q42435.xml");

# Load xml into memory / into a tree
my $doc = $xs1->XMLin($html);

## What's needed to do what follows:
#print Dumper($doc);

# Retrieve the name
my $name = $doc->{entry}->{protein}->{recommendedName}->{fullName};
print "Protein has for name: $name \n";

# Retrieve the sort name
my $shortname = $doc->{entry}->{name};
print "Protein has for short-name: $shortname \n";

# Retrieve the different pathways available.
my @pathways;
my $cnt = 0;
while (my $comment = $doc->{entry}->{comment}[$cnt]){
    $cnt += 1;
    if ( $comment->{type} eq 'pathway'){
        push(@pathways, $comment->{text});
    }
}

# Iterate through all the pathway found and print the information
foreach my $pathway(@pathways) {
    print "Protein has pathway: $pathway\n";
}
